### ISSUE TITLE
[topgen] --top_ral switch also dumps a large number of RTL sources

### ISSUE THREAD

**Issue by sriyerg** [lowRISC/opentitan#7857]
`topgen` is written with the assumption that all sources are generated and checked into the repo. For DV (and likely SW as well) there are good reasons to generate sources in the scratch space only when the simulation is run, rather than checking in. 

The problem is in the `_process_top()` method, which not only creates and returns the data-structures that are used for the generation of RTL/DV/SW/doc sources, but it also unconditionally generates several top-specific RTL, HJSon and other collateral. Consequently, when `topgen` is invoked with the intent to only create the top level RAL model, these additional sources are also created in the scratch space. This is benign, but not ideal.

I tried to mask the generation of these sources (i.e. all `generate_*()` function calls) by putting them inside `if not args.top_ral` clause, but seems like there is some kind of dependency on those functions too (especially, `generate_clkmgr()`) to successfully return from the `_process_top()` invocation. The following errors get thrown:
```
ERROR:  CLKMGR mismatched number of reset ports and nets
ERROR:  CLKMGR Following reset ports do not exist:
ERROR: rst_main_ni
ERROR: rst_io_ni
ERROR: rst_usb_ni
ERROR: rst_io_div2_ni
ERROR: rst_io_div4_ni
ERROR: module clock/reset checking failed
ERROR: xbar clock/reset checking failed
Error occured while validating top.hjson
```

Filing an issue to get this properly refactored later.

**Comment by sriyerg** [comment]
It will also be great to be able to invoke `topgen` just to query some information (which can be used in other scripts), without having a bunch of files being generated.

**Comment by sriyerg** [comment]
Looks like the issue is having to generate HJson data itself from the template on which the top depends on - perhaps for these IPs we only render the mako template (for the HJSon) in `_process_top` as a string first? We can convert that string into HJSon data directly without extracting from the generated file (and we maintain these IP specific HJSon data in the `topcfg` data structure so that we can come back to it later when we want to generate those files). This will allow us to move all of the `generate_*()` calls to outside `_process_top()`.

**Comment by tjaychen** [comment]
see #2083, this isn't quite so straightforward.

It isn't just that we need to generate hjson data that the top depends on, it's also that generation of hjson between certain modules have a dependency.  So process_top needs to be called a couple of times, so that the first generation creates the proper hjson (for some modules) and the second pass creates the other module hjsons for those that have a dependency.  It's all very very messy and we need to figure it out how to better handle it one day (even this multi-pass doesn't completely work).

I THINK what you're getting at though, is to not generate the rtl (and other) collateral if we don't need it? If that's the case, i think it would be fine for `_process_top` to generate only the hjson first (if nothing else...this keeps the parsing method consistent between the 1st and 2nd runs), but hold off on generating everything else later.

Unless, is an hjson in string format can be parsed just like a normal hjson?

**Comment by sriyerg** [comment]
Yes - mako can return a string of what it processed without actually having to write the file. And you can convert that string to HJson data structure. So essentially we could completely postpone the generation of all sources.

**Comment by tjaychen** [comment]
is there any reason to postpone the hjson generation also?
the others i can see a rationale, but i'm not sure what the reason to
skipping hjson also.  Since you need that for the DV collateral anyways.

**Comment by sriyerg** [comment]
I provide the scratch path to generate the RAL sources, and not the repo_top path. So what happens is it re-creates the whole directory structure in the scratch area to generate the HJSon file at the right place. But this is a duplicate of what is already checked-in. `topgen` has no idea that this is duplicate, which is not used by anything else. 

One solution would be to make `topgen` understand where the repo_top is (which would be separate from `--outdir`), so that it always creates these sources relative to repo_top rather the `--outdir`. But this may make things even more confusing.

**Comment by tjaychen** [comment]
o i see. so in this case, you are assuming that all the collateral is
already in place, and we just want to go ahead and directly generate the dv
collateral, is that right?

**Comment by sriyerg** [comment]
Right, because we are not checking in the DV collateral.

**Comment by tjaychen** [comment]
yeah that makes sense.
So when we pass in the ral_switch, i guess really we should skip all the
data structure generation and skip directly ahead?
That seems like a very reasonable thing to do.  We'd probably just need to
refactor the code a bit.

**Comment by sriyerg** [comment]
Hmm, to be precise, when we pass the ral switch, we need `_process_top` to generate the full topcfg data structure (which our code heavily relies on), but not create any of the files.

**Comment by tjaychen** [comment]
i can't remember now, but isn't topcfg mostly top_earlgrey.gen.hjson?
maybe if we can try to make those two equivalent, that would solve a lot of
problems.
